lib/report.js

```diff
@@ -158,7 +160,152 @@ class ConciseReporter extends Reporter {
   }
 }
 
+class TapReporter extends Reporter {
+  constructor(options) {
+    super(options);
+    this.successful = 0;
+    this.failed = 0;
+    this.skipped = 0;
+    this.failedTests = [];
+    this.testNumeration = [1];
+    this.lastIndex = 0;
+    this.message = '';
+    this.firstFailure = true;
+    this.got = [];
+    this.expected = [];
+  }
+
+  error(test, error) {
+    // remove one success because 'error' may only be received after end
+    // though the test may have already failed so we need a way to check
+    // if we should do this, i.e. store map of successful tests
+    // this.successes--;
+    // test.todo ? this.skipped++ : this.failures++;
+    // TODO(lundibundi): add test.metadata.filename here too once available
+    this.listFailure(
+      test,
+      { message: `Failure after end: ${error ? error.message : 'no message'}` },
+      ` Test failed: ${test.caption}`
+    );
+  }
+
+  parseTestResults(test) {
+    this.testNumeration[this.lastIndex]++;
+    if (!test.success) {
+      if (!test.todo) {
+        this.failedTests.push(this.testNumeration[this.lastIndex] - 1);
+      }
+      this.listFailure(test, true);
+      this.got.forEach(el => {
+        this.message += `${pad(6)}- ${el}\n`;
+      });
+      this.message += `${pad(4)}expect: \n`;
+      this.expected.forEach(el => {
+        this.message += `${pad(6)}- ${el}\n`;
+      });
+      this.message += '  ...\n';
+      this.got = [];
+      this.expected = [];
+      console.log(this.message);
+    } else {
+      this.testNumeration.push(0);
+      this.lastIndex++;
+      this.message = pad(this.lastIndex * 2);
+      test.results.forEach(res => {
+        if (res.success) {
+          if (res.type === 'subtest') {
+            const t = res.test;
+            console.log(`${this.message}# Subtest: ${t.caption}`);
+            this.parseTestResults(t);
+            this.message = pad(this.lastIndex * 2);
+            const prefix = t.success ? 'ok' : 'not ok';
+            let message = `${this.message}${prefix} ${
+              this.testNumeration[this.lastIndex]
+            }`;
+            message += ` - ${t.caption}`;
+            console.log(message);
+          } else {
+            this.listSuccess(res);
+          }
+        }
+      });
+      console.log(`${this.message}1..${this.testNumeration[this.lastIndex]}`);
+      this.lastIndex--;
+      this.testNumeration.pop();
+    }
+    if (!this.lastIndex) this.message = '\n';
+  }
+
+  listSuccess(test) {
+    const testPrefix = test.success ? 'ok' : 'not ok';
```

Comment:
The function is called `listSuccess` but it still checks `test.success` this looks incorrect to me.

Reply:
Why changing the name of function to checkSuccess can't resolve the problem ?

Reply:
This check is not needed as this method is only called when 'test' is successful.
Therefore testPrefix should always be 'ok' and method be renamed to listSuccess in compliance with listFailure.

Reply:
This check is not needed as this method is only called when 'test' is successful.
Therefore testPrefix should always be 'ok' and method be renamed to listSuccess in compliance with listFailure.